infra.bs

```diff
@@ -282,8 +283,106 @@ a simple wrapper around another one.)
+ optional boolean <var ignore>exceptionsEnabled</var> (default false):
+
+ <ol class=brief>
+  <li>&hellip;
+  <li>If <var ignore>navigationType</var> was given, then do something with
+  <var ignore>navigationType</var>.
+  <li>&hellip;
+ </ol>
+</div>
+
+<p>To call algorithms with such optional positional parameters, the optional argument values can be
+omitted, but only the trailing ones.
+
+<div class=example id=example-algorithm-optional-positional-params-calling>
+ <p>Call sites to the previous example's algorithm would look like one of:
+
+ <ul class=brief>
+  <li><a href=#example-navigate-algo-positional>Navigate</a> to <var ignore>resource</var>.
+
+  <li><a href=#example-navigate-algo-positional>Navigate</a> to <var ignore>resource</var> with
+  "<code>form submission</code>".
+
+  <li><a href=#example-navigate-algo-positional>Navigate</a> to <var ignore>resource</var> with
+  "<code>form submission</code>" and true.
+ </ul>
+
+ <p>But, there would be no way to supply a non-default value for the third
+ (<var ignore>exceptionsEnabled</var>) argument, while leaving the second
+ (<var ignore>navigationType</var>) argument as not-given.
+</div>
+
+<p>Optional named parameters, instead of positional ones, can be used to increase clarity and
+flexibility at the call site. Such parameters are marked up as both variables and definitions, and
+linked to from their call sites.
+
+<div class=example id=example-algorithm-optional-named-params>
+ <p class=allow-2119>To <dfn ignore id=example-navigate-algo-named>navigate</dfn> given a
+ resource <var ignore>resource</var>, an optional string
+ <dfn ignore id=example-navigate-algo-navigationType><var ignore>navigationType</var></dfn>, and an
+ optional boolean
+ <dfn ignore id=example-navigate-algo-exceptionsEnabled><var ignore>exceptionsEnabled</var></dfn>
+ (default false):
+
+ <ol class=brief>
+  <li>&hellip;
+  <li>If <var ignore>navigationType</var> was given, then do something with
+  <var ignore>navigationType</var>.
+  <li>&hellip;
+ </ol>
+
+ <p>Call sites would then look like one of:
+
+ <ul class=brief>
+  <li><a href=#example-navigate-algo-named>Navigate</a> to <var ignore>resource</var>.
+
+  <li><a href=#example-navigate-algo-named>Navigate</a> to <var ignore>resource</var> with
+  <a href=#example-navigate-algo-navigationType><var ignore>navigationType</var></a> set to
+  "<code>form submission</code>".
+
+  <li><a href=#example-navigate-algo-named>Navigate</a> to <var ignore>resource</var> with
+  <a href=#example-navigate-algo-exceptionsEnabled><var ignore>exceptionsEnabled</var></a> set to
+  true.
+
+  <li><a href=#example-navigate-algo-named>Navigate</a> to <var ignore>resource</var> with
+  <a href=#example-navigate-algo-navigationType><var ignore>navigationType</var></a> set to
+  "<code>form submission</code>" and
+  <a href=#example-navigate-algo-exceptionsEnabled><var ignore>exceptionsEnabled</var></a> set to
+  true.
```

Comment:
If booleans default to false, should referencing them here set them to true? It seems to read well:

Navigate to resource with exceptionsEnabled.

Reply:
Oh that's interesting. We should perhaps also allow/require omitting "(default false)" for booleans and basically have built-in support for that type, or is that too much?

Reply:
I'd rather not special-case booleans in that way.

Reply:
In particular the benefit of "(default false)" is that it's obvious what the default is. And with "exceptionsEnabled set to true", it's obvious what the value of exceptionsEnabled is.

If you remove either of those, then the reader needs to know the Infra special-casing rules pretty intimately, otherwise they're left confused. "Why did this call site not pass exceptionsEnabled? Unlike navigationType, it has no default, so how is that possible?" "This call site said with exceptionsEnabled, but what is exceptionsEnabled actually set to? Are they trying to reference a nearby exceptionsEnabled variable's value?"

Reply:
Yeah, I guess it's reasonable to require it for that reason, but we could require it to always be "(default false)" and forbid "(default true)".

Reply:
I'd be on board with that particular special-casing :).